### DISCUSSION TITLE
Cycle basis for directed graph?

### DISCUSSION THREAD

**Discussion by kimin0402** [networkx/networkx#5640] [Q&A] [answered]
Hi, thank you for such a wonderful package. 

Sorry if my question is from my ignorance of graph theory. I don't have a strong mathematical background, but I know linear algebra and the concept of basis for the cycle space. 

I was wondering if it is possible to get cycle basis for a directed graph. Simple_cycles function seems to return unique list of cycles in a directed graph, but I'm not sure they are the basis of the cycle space. (Because we can estimate the number of cycle basis, and simple cycle seems to produce cycles beyond the number of expected cycle basis. )

If getting cycle basis for a diG is not a good idea, is there any alternative to finding some sort of basis of diG? (The resulting vector space dose not have to be a cycle space. It could be even better if I could get basis for maybe, the entire graph space or if possible, some sort of simple path space. Maybe possibly extracting filament basis and cycle basis? I'm not so sure)

Thank you in advance.

**Comment by dschult** [comment, marked as answer]
This article on [directed cycle bases](https://www.sciencedirect.com/science/article/pii/S0166218X06003052.) looks like a good place to start looking at all the possibilities. That paper defines the "oriented cycle basis" of a directed graph as a set of cycles of the undirected version of the graph along with the orientation of each edge in the cycle. So, the cycle includes edges that are in directions counter to the order of the nodes in the cycle.

In NetworkX, the cycle_basis function only works for undirected graphs, and simple_cycles will give far more than a cycle basis. I'm guessing some reading about notions of directed cycle spaces is in order because which definition you want likely depends what question you are investigating.

**Reply by kimin0402** [reply]
Thank you for your fast reply

So from my understanding, it seems like the notion of cycle basis for undirected graphs cannot be straightly applied to the directed graphs, and the closest equivalent version of cycle basis for directed graph is possible "oriented cycle basis" defined in the paper. Is it correct?
